### ISSUE TITLE
Implement #1389

### ISSUE THREAD

**Issue by somedadaism** [facebookresearch/hydra#1939]
# 🚀 Feature Request

I would like to ask if it is possible to implement the functionality asked for in #1389

## Motivation

I believe the feature is extremely useful for ML applications. To be able to have list composition in the defaults list the way we already have dict composition would allow to be less redundant in the configuration by reusing config files more.

## Pitch

I know that #1389 had been closed as invalid but it seems that in omegaconf 2.1 the necessary prerequisites maybe have been implemented since the issue had been closed.
> > > OmegaConf 2.1 will support more powerful custom resolvers. It's _possible_ that they will enable generating as_list automatically from dict. (cc @odelalleau).
> > 
> > 
> > Not in my current implementation because resolvers systematically wrap their output in a `ValueNode`. However I suspect that if we used `_node_wrap()` instead then it should be possible. Not going to change it now though -- I'd rather keep it for later.
> 
> I think this should be possible now, with `oc.dict.keys`, see https://omegaconf.readthedocs.io/en/2.1_branch/custom_resolvers.html#oc-dict-keys-value

This could maybe be used to allow for list composition in the defaults list in the way originally asked for instead of using `os.dict.values` in a hacky fashion. 

**Are you willing to open a pull request?** 

Maybe if I can get sufficient guidance I would try.

**Comment by Jasha10** [comment]
Hi @somedadaism,
Thank you for your interest in Hydra and for taking the time to open this issue!

> This could maybe be used to allow for list composition in the defaults list in the way originally asked for...

The way originally asked for proposes that, with the following config, the db group would produce a ListConfig object containing each of the mysql and postgresql input configs:
```yaml
defaults:
  - base_config
  - db: 
    - mysql
    - postgresql
```
This will not be possible because there is already a different behavior implemented for nested lists within a defaults list (such as the nested list `- db: [mysql, postgresql]`): the mysql and postgresql configs are merged. See the Hydra docs on [Selecting multiple configs from a Config Group](https://hydra.cc/docs/patterns/select_multiple_configs_from_config_group/).

Admittedly I have not yet had a chance to look closely at later discussion in comments on the linked issue #1389.

**Comment by somedadaism** [comment]
Hi @Jasha10 thanks for your quick reply.  
I would like to clarify the functionality I search for. Please consider first the following (simplified) situation:
```
.
├── hydra_config
│   ├── callbacks
│   │   ├── my_callback_1.yaml
│   │   └── my_callback_2.yaml
│   └── config.yaml
└── main.py
```
With: 
```
# main.py
import hydra

@hydra.main(config_path="hydra_config", config_name="config_new.yaml")
def main(config):
    print(config)

if __name__ == "__main__":
    main()
```
```
# config.yaml
defaults:
  - callbacks:
    - my_callback_1
    - my_callback_2
```
What currently works is:
```
# my_callback_1.yaml
my_callback_1:
  my_param1: 1
```
```
# my_callback_2.yaml
my_callback_2:
  my_param2: 2
```
resulting in:
```
{'callbacks': {'my_callback_1': {'my_param': 1}, 'my_callback_2': {'my_param2': 2}}}
```
What I want to create as output is:
```
{'callbacks':  [{'my_param': 1}, {'my_param2': 2}]}
```
Now I understand that:
```
# my_callback_1.yaml
my_param1: 1
```
```
# my_callback_2.yaml
my_param2: 2
```
Can not work as it (correctly!) results in:
```
{'callbacks': {'my_param': 1, 'my_param2': 2}}
```
which is what I believe you also pointed out.
What is not understandable to me is that for an example:
```
# my_callback_1.yaml
- my_param1: 1
```
```
# my_callback_2.yaml
- my_param2: 2
```
will result in:
```
{'callbacks': [{'my_param2': 2}]}
```
and not in
```
{'callbacks':  [{'my_param': 1}, {'my_param2': 2}]}
```
I believe the ability to merge configs into lists is very important to avoid repeating oneself for ML applications of hydra, but there seems to be no way to merge them.  
(Imagine using the same 5 callbacks all the time in a list but being forced to type them out repeatedly for different experimental model runs - if you change one thing you have to change 5 config files if you are lucky enough to remember them all.). 
I actually believe it would be important enough to introduce new explicit notation to enable it.
Maybe something like:
```
# config.yaml
defaults:
  - callbacks:
  [
    - my_callback_1
    - my_callback_2
  ]
```

**Comment by omry** [comment]
Take a look at [oc.dict.values](https://omegaconf.readthedocs.io/en/2.1_branch/custom_resolvers.html#oc-dict-keys-value), it may be what you are looking for.
Basically, you can let Hydra compose a dict, and have a key that dynamically transform it to a list.

**Comment by somedadaism** [comment]
Hello @omry thank you for creating hydra and for taking a look.  
I am very aware of `oc.dict.values` actually the answer provided by you also proposed on StackOverflow and was linked by me in the original question.   
Sadly it is specifically what I don't want to do.  
My problems with this approach are the following:

1) It's not straight forward to read - it just feels hacky.
2) The approach forces us to create a dictionary in first place before casting to a list. This means the dictionary will also be part of the configuration. If now I just want to use `hydra.utils.instantiate` on my entire config it will create all objects created vie `_target_` twice, once for the dict and then for the list. For some objects, e.g. some of my code for W&B this is a problem.
I know multiple ways to hack around this but I would rather invest time to get a cleaner solution in `hydra` itself.
3) I feel the current defaults feel inconsistent. When selecting multiple configs we replace the respective line in the defaults with the content of the referenced file and then merge if necessary. If this would apply for configs with lists too in the situation:
```
# my_callback_1.yaml
- my_param1: 1
```
```
# my_callback_2.yaml
- my_param2: 2
```
Then:  
```
# config.yaml
defaults:
  - callbacks:
    - my_callback_1
    - my_callback_2
```
Would be text substituted to:
```
callbacks:
  - my_param1: 1
  - my_param2: 2
```
Just as in the situation:
```
# my_callback_1.yaml
my_callback_1:
  my_param1: 1
```
```
# my_callback_2.yaml
my_callback_2:
  my_param2: 2
```
The defaults list in:
```
# config.yaml
defaults:
  - callbacks:
    - my_callback_1
    - my_callback_2
```
Is text substituted to:
```
callbacks:
  my_callback_1:
    my_param1: 1
  my_callback_2:
    my_param2: 2
```
I know this is not how it works but I believe it would be the more natural and consistent way.
I have discussed on this with multiple of my coworkers and the ability to merge lists in the defaults list is missed by us very much. Is there really no less hacky way than `oc.dict.values`?

**Comment by somedadaism** [comment]
@omry @Jasha10 Any updates on this?

**Comment by Jasha10** [comment]
The current behavior is probably due to the implementation of `OmegaConf.merge`:
```python
>>> from omegaconf import OmegaConf
>>> OmegaConf.merge([1,2], [3,4])
[3, 4]
>>> OmegaConf.merge({"a": [1,2], "b": [3,4]}, {"b": [5,6], "c": [7,8]})
{'a': [1, 2], 'b': [5, 6], 'c': [7, 8]}
```
mappings on the left are updated by mappings on the right; sequences on the left are overwritten by sequences on the right.

I agree with you that, in the specific case of "Selecting multiple configs from a Config Group" in the defaults list, it would be more useful to concatenate ListConfigs instead of overwriting them (as you have outlined above). The proposed change would technically be breaking, but I doubt anyone is relying on the old behavior.

This being said, I'm not sure that special-casing the "selecting multiple configs from a config group" behavior is the most general solution to the use-case you've highlighted. It still feels like a hack that `my_callback_1.yaml` has to be written with a leading dash:
```yaml
# my_callback_1.yaml
- my_param1: 1
```
Painfully, we are prevented from using a defaults list in `my_callback_1.yaml` if the file is written as a yaml sequence (as above).

You are trying to achieve the following result:
`{'callbacks':  [{'my_param1': 1}, {'my_param2': 2}]}`.
Inspired by #1547, I propose the following:
```yaml
# config.yaml
defaults:
  - append callbacks: my_callback_1
  - append callbacks: my_callback_2
```
Here an "append" keyword would signify that the `callbacks` package should be a ListConfig, and that the selected `my_callback_{1,2}` config should be appended to that list. These configs `my_callback_1.yaml` and `my_callback_2.yaml` would be written as yaml mappings, possibly having defaults lists of their own:
```yaml
# my_callback_1.yaml
defaults: ...  # possible defaults list
my_param1: 1  # no leading dash
```

**Comment by somedadaism** [comment]
I think the proposed syntax is very good.   
I would take a look at contributing but only have limited time.   
Where could I get hints on how to achieve this?

**Comment by Jasha10** [comment]
Unfortunately, I fear that the implementation of this feature might not be straight-forward; it will require adding a new keyword to the [`ConfigRepository.Keywords`](https://github.com/facebookresearch/hydra/blob/7751c9bcae17f9819bf466cf900e68c13e1296c9/hydra/_internal/config_repository.py#L302) class and modifying the logic in [`defaults_list.py`](https://github.com/facebookresearch/hydra/blob/7751c9bcae17f9819bf466cf900e68c13e1296c9/hydra/_internal/defaults_list.py).

**Comment by Jasha10** [comment]
Cross-link to #1547 which is related.

**Comment by tangbinh** [comment]
@Jasha10 @jieru-hu I agree with @somedadaism that this feature is very much needed for ML applications. We often pass a list of callbacks to a trainer, and being able to compose such a list via the command line is necessary. Given the huge number of combinations one can derive from a list (e.g. for 5 callbacks, there are 32 combinations), it's impossible to have a YAML for each combination.

@Jasha10 If it's not easy to implement this feature, is there an acceptable workaround in the meantime? For example, I would like to specify a default list of callbacks in a primary config and couldn't figure out a way to do it yet.

```yaml
# cb1.yaml
_target_: torch.nn.Identity
arg1: 1
```

```yaml
# cb2.yaml
_target_: torch.nn.Identity
arg2: 2
```

```yaml
# config.yaml
defaults:
  - callbacks:
    # what to put here?
```

The desired output looks like
```yaml
callbacks:
- _target_: torch.nn.Identity
  arg1: 1
- _target_: torch.nn.Identity
  arg2: 2
```

**Comment by Jasha10** [comment]
@tangbinh, see Omry's [comment](https://github.com/facebookresearch/hydra/issues/1939#issuecomment-1003108756) above.

Basically, you should use the defaults list to compose a dictionary whose values are the callbacks you want. You then can use the `oc.dict.values` resolver to get a list of callbacks
```yaml
# config.yaml
defaults:
  - callbacks@_callback_dict.cb1: cb1
  - callbacks@_callback_dict.cb2: cb2
  - _self_

callbacks: ${oc.dict.values:_callback_dict}
```
```python
# my_app.py
import hydra
from omegaconf import OmegaConf

@hydra.main(config_path=".", config_name="config")
def app(cfg):
    OmegaConf.resolve(cfg)
    del cfg._callback_dict
    print(OmegaConf.to_yaml(cfg))

if __name__ == "__main__":
    app()
```
```text
$ python my_app.py
callbacks:
- _target_: torch.nn.Identity
  arg1: 1
- _target_: torch.nn.Identity
  arg2: 2
```

**Comment by tangbinh** [comment]
@Jasha10 Thank you for working out the example. It's certainly helpful, although I think we all agree that invoking `@_callback_dict` and `${oc.dict.values:_callback_dict}` reduces readability significantly. Can you also work out how I can append another callback to the list via the command line?

If possible, would we be able to prioritize this list composition feature a bit? I believe ML users would run into this issue quite often.

**Comment by Jasha10** [comment]
> If possible, would we be able to prioritize this list composition feature a bit? 

Thanks for the nudge :)
I agree completely that improved support for list composition is an important feature for user experience.

> Can you also work out how I can append another callback to the list via the command line?

Here's the trick:
```text
$ python my_app.py +callbacks@_callback_dict.cb3=cb3
callbacks:
- _target_: torch.nn.Identity
  arg1: 1
- _target_: torch.nn.Identity
  arg1: 2
- _target_: torch.nn.Identity
  arg1: 3
```
The plus symbol is for [appending to the defaults list](https://hydra.cc/docs/advanced/override_grammar/basic/#modifying-the-defaults-list).

**Comment by Yevgnen** [comment]
@Jasha10 

```yaml
# config.yaml
defaults:
  - callbacks@_callback_dict.cb1: cb1
  - callbacks@_callback_dict.cb2: cb2
  - _self_

callbacks: ${oc.dict.values:_callback_dict}
```

Hi, in this example, is there a way to avoid writing `callbacks@_callback_dict.cb{x}: cb{x}` for every `{x}` so that we can include all items under `callback/` at once?

**Comment by Jasha10** [comment]
I am starting to think that introducing a new keyword would not be as flexible as using bracket notation.
```yaml
defaults:
  - callbacks_list[]: callback1
```
This would result in a the following composed DictConfig:
```python
{"callbacks_list": [<callback1>]}
```
My reason for preferring bracket notation is that the bracket notation can express nested lists:
```
defaults:
  - models[].callbacks_list[].param: param1
```
```python
{"models": [{"callbacks_list": [{"param": <param1>}]}]}
```
This idea is not fully fleshed out -- some syntax would be required for distinguishing between different list operations (append/extend/insert/...).

**Comment by SsnL** [comment]
The workaround here doesn't work nicely with structured config. Unless the config also have a `_callback_dict` entry, OmegaConf will complain in merging. :(

**Comment by TesnimHadhri** [comment]
> The workaround here doesn't work nicely with structured config. Unless the config also have a `_callback_dict` entry, OmegaConf will complain in merging. :(

Do you mind explaining the workaround you used with structured config?

**Comment by SsnL** [comment]
@TesnimHadhri I eventually did not use it, but I remember having a `_callback_dict` field with dictionary type helped. but YMMV

**Comment by TesnimHadhri** [comment]
> @TesnimHadhri I eventually did not use it, but I remember having a `_callback_dict` field with dictionary type helped. but YMMV

I see, thanks for the quick answer!

**Comment by libokj** [comment]
@Jasha10 Hi, is it possible for you to provide a brief update on the status of the implementation of this feature? Should we look into it as you commented earlier?

> Unfortunately, I fear that the implementation of this feature might not be straight-forward; it will require adding a new keyword to the [`ConfigRepository.Keywords`](https://github.com/facebookresearch/hydra/blob/7751c9bcae17f9819bf466cf900e68c13e1296c9/hydra/_internal/config_repository.py#L302) class and modifying the logic in [`defaults_list.py`](https://github.com/facebookresearch/hydra/blob/7751c9bcae17f9819bf466cf900e68c13e1296c9/hydra/_internal/defaults_list.py).

**Comment by anton-bushuiev** [comment]
Note for the ones still seeking for a better solution: [Lightning-Hydra-Template](https://github.com/ashleve/lightning-hydra-template/tree/main/configs/callbacks) implements another workaround which is however still based on a dictionary with redundant keys. The difference between the [solution](https://github.com/facebookresearch/hydra/issues/1939#issuecomment-1035395006) suggested by @Jasha10 above is that the dict is parsed internally by a custom instantiator rather than relying on `oc.dict.values` which increases readability significantly but loses succinctness.

**Comment by libokj** [comment]
> Note for the ones still seeking for a better solution: [Lightning-Hydra-Template](https://github.com/ashleve/lightning-hydra-template/tree/main/configs/callbacks) implements another workaround which is however still based on a dictionary with redundant keys. The difference between the [solution](https://github.com/facebookresearch/hydra/issues/1939#issuecomment-1035395006) suggested by @Jasha10 above is that the dict is parsed internally by a custom instantiator rather than relying on `oc.dict.values` which increases readability significantly but loses succinctness.

That's the workaround I am resorting to as well. Still hope this feature can be implemented.

**Comment by alex-hh** [comment]
I also would find this extremely useful: anywhere a functional style is used it is useful to specify an ordered list of items.

As well as callbacks, pipelines of preprocessing transformations seem like a natural use case.

Do the current workarounds preserve the order of the composed items? This is critical in such cases but I'm unclear whether the fact that dictionaries are involved in intermediate steps means that the order is not preserved

**Comment by dylandronnier** [comment]
Crucial feature for me as well...

I can't make the workaround from @Jasha10 work with Structured Config. Do you know if the solution from [https://github.com/ashleve/lightning-hydra-template](Lightning-hydra-template) work with Structured Config?

**Comment by cy94** [comment]
+1, this would be very useful to configure, for eg. multiple datasets for train/val, multiple loss functions, multiple transforms and so on

**Comment by omry** [comment]
Thanks for the detailed discussion here, and sorry for the long delay.

This ultimately depends on better list merge semantics in OmegaConf rather than a Hydra-specific list composition mechanism. I am tracking the broader design, including Hydra's need to compose config-group selections into ordered lists, in https://github.com/omry/omegaconf/issues/1312.

I closed this issue in favor of that upstream tracker. Once the OmegaConf behavior is designed, Hydra can integrate it into config composition.
